etcddb/etcddb_client.go

```diff
@@ -262,59 +268,240 @@ func (client *EtcdClient) Transaction(compare []EtcdKeyValue, swap []EtcdKeyValu
+		if err != nil {
+			return false, err
+		}
+		newValues, err := request.Update(oldValues)
+		if err != nil {
+			return false, err
+		}
+		if ok, err = client.CompleteTransaction(transaction, newValues); err != nil {
+			return false, err
+		}
+		if ok {
+			return true, nil
+		}
+		if request.RetryTillSuccessOrError {
+			continue
+		}
+	}
+}
+
+//If there are no Old values in the transaction, to compare, then this method
+//can be used to write in the new values , if the key does not exist then put it in a transaction
+
+func (client *EtcdClient) CompleteTransaction(_transaction escrow.Transaction, update []*escrow.KeyValueData) (
+	ok bool, err error) {
+
+	var transaction *etcdTransaction = _transaction.(*etcdTransaction)
 
-	// TODO: implement it using etcdv3.KV.Txn(ctx).If(...).Then(...).Commit() as in CompareAndSwap()
-	mu := concurrency.NewMutex(session, key)
-	err = mu.Lock(ctx)
+	ctx, cancel := context.WithTimeout(context.Background(), client.timeout*time.Second*100)
+	defer cancel()
+	defer ctx.Done()
+
+	txn := client.etcdv3.KV.Txn(ctx)
+	var txnResp *clientv3.TxnResponse
+	if txn, err = client.buildIf(txn, transaction); err != nil {
+		return false, err
+	}
+	if txn, err = client.buildThenOperations(txn, update); err != nil {
+		return false, err
+	}
+	if txn, err = client.buildElseOperations(txn, GetKeysFromKeyValueData(update)); err != nil {
+		return false, err
+	}
+	txnResp, err = txn.Commit()
 
 	if err != nil {
-		log.WithError(err).Error("Unable to lock mutex")
-		return
+		return false, err
+	}
+
+	if txnResp == nil {
+		return false, fmt.Errorf("transaction response is nil")
+	}
+	if txnResp.Succeeded {
+		return true, nil
 	}
 
-	defer mu.Unlock(context.Background())
+	var latestValues []*keyValueVersion
+	if latestValues, err = client.checkTxnResponse(txnResp); err != nil {
+		return false, err
+	}
 
-	response, err := etcdv3.Get(ctx, key)
+	transaction.ConditionValues = latestValues
+	//succeeded is set to true if the compare evaluated to true or false otherwise.
+	return txnResp.Succeeded, nil
+}
 
-	if err != nil {
-		log.WithError(err).Error("Unable to get value")
-		return
+func GetKeysFromKeyValueData(update []*escrow.KeyValueData) []string {
+	keys := make([]string, len(update))
+	for i, key := range update {
+		keys[i] = key.Key
 	}
-	if response.Count != 0 {
-		return
+	return keys
+}
+
+func (client *EtcdClient) buildIf(txn clientv3.Txn, transaction *etcdTransaction) (clientv3.Txn, error) {
+	if len(transaction.ConditionValues) == 0 {
+		return txn.If(client.alwaysTrueCompare()), nil
 	}
+	cmps := make([]clientv3.Cmp, len(transaction.ConditionValues))
 
-	_, err = etcdv3.Put(ctx, key, value)
+	for i, cmp := range transaction.ConditionValues {
+		cmps[i] = clientv3.Compare(clientv3.ModRevision(cmp.Key), "=", cmp.Version)
+	}
+	return txn.If(cmps...), nil
+}
 
-	if err != nil {
-		log.WithError(err).Error("Unable to put value")
-		return
+func (client *EtcdClient) alwaysTrueCompare() clientv3.Cmp {
```

Comment:
@vsbogd , need your help to see if there is a elegant way here , was looking a reading all values in a single transaction , we use specific keys now

Reply:
@anandrgitnirman , I am not sure I understand the question. But I would simplify code and remove all this `If/Then` stuff. When I suggested it I thought we can implement it on storage API level to give user the flexibility of forming some expressions which could be translated into ETCD transaction expressions. Eventually we agreed that it is not simple (if even possible) in ETCD API and moved to an API where user can execute arbitrary code but it should provide the set of keys which he expects not to be changed while new state is calculated. So adding some custom `If/Then` at the moment doesn't add any value. 

I would propose to remove `If/Then` and instead write plain method which checks that condition key versions were not changed and applies new values to the database. My view it should be possible using the same technique which were used in previous `Transaction` method implementation: https://github.com/singnet/snet-daemon/blob/1a5298472c615f6b155317c7f540427b20360931/etcddb/etcddb_client.go#L237-L246

Do you see some difficulties here which I missed?

Reply:
hi @vsbogd ,

when the IF condition Fails
  Reason 1: , the data was modified in a different transaction and hence  we need to read the Values again from the  else condition 
 Reason 2: The Key itself is not there , => we need to still persist ( so PutIfAbsent ) 

Let me brain storm this a little more and come back

Reply:
> Reason 1: , the data was modified in a different transaction and hence we need to read the Values again from the else condition

Ok, do you mean that if `If` condition fails you would like to get new data for conditional values in same transaction to make highlevel transaction more effective? 

> Reason 2: The Key itself is not there , => we need to still persist ( so PutIfAbsent )

To check that key exist or not exist you can check `CreateRevision` according to https://github.com/etcd-io/etcd/issues/7115#issuecomment-271130979

Reply:
> Reason 1: , the data was modified in a different transaction and hence we need to read the Values again from the else condition

To implement this why not add `OpGet` into `Else` branch?

Reply:
@vsbogd , addressed these , could you check now ?

Reply:
Please see my suggestion: https://github.com/anandrgitnirman/snet-daemon/pull/47